cilium, bpf: make global vs local CT table configurable

borkmann opened the pull request:
Use a single huge global connection tracking table by default, and
have also the option to switch to container local one instead as we
have prior to this patch.

The switch can even happen during runtime per reconfiguration via:

  cilium endpoint config <ep-ID> ConntrackLocal=false  // global table
  cilium endpoint config <ep-ID> ConntrackLocal=true   // local table

Dumping the global table works as usual:

  cilium bpf ct list global    // dump global table
  cilium bpf ct list <ep-ID>   // dump local table

It's also listed as a config option via cli, f.e.:

  # cilium config
  Conntrack                Enabled
  ConntrackAccounting      Enabled
  ConntrackLocal           Enabled
  DropNotification         Enabled
  NAT46                    Enabled
  Policy                   Enabled
  PolicyTracing            Disabled

I changed the 01-ct tests to include local/global table tests and
they were all passing.

Signed-off-by: Daniel Borkmann <daniel@cilium.io>
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/cilium/cilium/pull/443%23discussion_r107949091%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23discussion_r107949642%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23issuecomment-289079033%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23issuecomment-289079342%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23issuecomment-289079469%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23discussion_r107953108%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23issuecomment-289083804%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23discussion_r107954562%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23discussion_r107955489%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23discussion_r107956658%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23discussion_r107956746%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23discussion_r107958431%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23issuecomment-289106734%22%2C%20%22https%3A//github.com/cilium/cilium/pull/443%23issuecomment-289144483%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/443%23issuecomment-289079033%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22The%20CT%20key%20is%20currently%20defined%20as%20%3A%5Cr%5Cn%60%60%60%5Cr%5Cntype%20CtKey4%20struct%20%7B%5Cr%5Cn%20%20%20%20%20%20%20%20addr%20%20%20%20types.IPv4%5Cr%5Cn%20%20%20%20%20%20%20%20sport%20%20%20uint16%5Cr%5Cn%20%20%20%20%20%20%20%20dport%20%20%20uint16%5Cr%5Cn%20%20%20%20%20%20%20%20nexthdr%20u8proto.U8proto%5Cr%5Cn%20%20%20%20%20%20%20%20flags%20%20%20uint8%5Cr%5Cn%7D%5Cr%5Cn%60%60%60%5Cr%5Cn%5Cr%5CnThere%20is%20only%20one%20address%20element%20in%20the%20key%20so%20far%2C%20this%20was%20possible%20because%20the%20table%20was%20per%20endpoint%20which%20means%20either%20the%20source%20or%20destination%20address%20is%20always%20the%20same.%20The%20primary%20motivation%20for%20this%20was%20to%20safe%20space%20in%20the%20key.%20As%20we%20move%20to%20the%20global%20table%2C%20both%20addresses%20need%20to%20be%20part%20of%20the%20key%20obviously.%22%2C%20%22created_at%22%3A%20%222017-03-24T16%3A52%3A01Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%2C%20%7B%22body%22%3A%20%22FYI%2C%20cilium%20should%20already%20remove%20stale%20ct%20maps%20on%20startup%2C%20see%20function%20%60staleMapWalker%28%29%60%22%2C%20%22created_at%22%3A%20%222017-03-24T16%3A53%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%2C%20%7B%22body%22%3A%20%22Okay%2C%20good%20point.%20I%27ll%20see%20to%20address%20that.%22%2C%20%22created_at%22%3A%20%222017-03-24T16%3A53%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/677393%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/borkmann%22%7D%7D%2C%20%7B%22body%22%3A%20%22Daniel%2C%20I%20am%20working%20on%20the%20tuple%20change%20part%20so%20it%20works%20for%20both%204%20tuple%20/%20local%20and%205%20tuple%20/%20global.%20I%20can%20send%20a%20pull%20request%20by%20tomorrow.%20%22%2C%20%22created_at%22%3A%20%222017-03-24T17%3A09%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/725475%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mchalla%22%7D%7D%2C%20%7B%22body%22%3A%20%22Looks%20good%20to%20me.%22%2C%20%22created_at%22%3A%20%222017-03-24T18%3A25%3A39Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/725475%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mchalla%22%7D%7D%2C%20%7B%22body%22%3A%20%22Repushed%3A%20fixed%20help%20message%20and%20added%20stale%20global%20map%20detection%20%26%20removal%20during%20daemon%20init.%22%2C%20%22created_at%22%3A%20%222017-03-24T21%3A13%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/677393%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/borkmann%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20b3540f1ad5c58546cd169149a9c95467168dbd90%20tests/01-ct.sh%207%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/443%23discussion_r107949091%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Agreed.%20reconfig%20should%20remove%20old%20stale%20maps.%20Cilium%20should%20also%20remove%20stale%20maps%20on%20startup.%22%2C%20%22created_at%22%3A%20%222017-03-24T16%3A48%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%2C%20%7B%22body%22%3A%20%22Potentially%20cilium%20could%20even%20do%20best-effort%20and%20migrate%20existing%20connection%20from%20the%20global%20to%20the%20local%20table.%22%2C%20%22created_at%22%3A%20%222017-03-24T16%3A50%3A58Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/677393%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/borkmann%22%7D%7D%2C%20%7B%22body%22%3A%20%22%5Cud83d%5Cudc4d%20%22%2C%20%22created_at%22%3A%20%222017-03-24T17%3A07%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20tests/01-ct.sh%3AL198-233%22%7D%2C%20%22Pull%20b3540f1ad5c58546cd169149a9c95467168dbd90%20pkg/endpoint/endpoint.go%2014%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/443%23discussion_r107954562%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22While%20I%20was%20reading%20the%20code%20I%20was%20asking%20my%20self%20%5C%22What%20is%20local%3F%20Do%20we%20have%20a%20non-local%20CT%3F%5C%22%20Now%20I%20get%20it%2C%20can%20I%20purpose%20to%20change%20the%20name%20to%20%60Use%20container%20dedicated%20connection%20tracking..%60%3F%20It%27s%20just%20a%20thought%22%2C%20%22created_at%22%3A%20%222017-03-24T17%3A14%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/5714066%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aanm%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20think%20we%20are%20using%20the%20word%20endpoint%20pretty%20much%20everywhere%20as%20Cilium%20may%20be%20extended%20to%20non%20container%20specific%20use%20cases.%20Let%27s%20be%20consistent.%22%2C%20%22created_at%22%3A%20%222017-03-24T17%3A18%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%2C%20%7B%22body%22%3A%20%22Makes%20sense%2C%20sure.%22%2C%20%22created_at%22%3A%20%222017-03-24T17%3A23%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/677393%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/borkmann%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40tgraf%20oh%20I%20was%20asking%20the%20%60local%60%20vs%20%60dedicated%60.%20The%20word%20container%20was%20already%20there.%22%2C%20%22created_at%22%3A%20%222017-03-24T17%3A23%3A41Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/5714066%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aanm%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%27m%20using%20%60Use%20endpoint%20dedicated%20tracking%20table%20instead%20of%20global%20one%60%20now.%22%2C%20%22created_at%22%3A%20%222017-03-24T17%3A31%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/677393%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/borkmann%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20pkg/endpoint/endpoint.go%3AL83-95%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [ ] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/cilium/cilium/pull/443#issuecomment-289079033'>General Comment</a></b>
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> The CT key is currently defined as :
```
type CtKey4 struct {
addr    types.IPv4
sport   uint16
dport   uint16
nexthdr u8proto.U8proto
flags   uint8
}
```
There is only one address element in the key so far, this was possible because the table was per endpoint which means either the source or destination address is always the same. The primary motivation for this was to safe space in the key. As we move to the global table, both addresses need to be part of the key obviously.
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> FYI, cilium should already remove stale ct maps on startup, see function `staleMapWalker()`
- <a href='https://github.com/borkmann'><img border=0 src='https://avatars2.githubusercontent.com/u/677393?v=3' height=16 width=16></a> Okay, good point. I'll see to address that.
- <a href='https://github.com/mchalla'><img border=0 src='https://avatars0.githubusercontent.com/u/725475?v=3' height=16 width=16></a> Daniel, I am working on the tuple change part so it works for both 4 tuple / local and 5 tuple / global. I can send a pull request by tomorrow.
- <a href='https://github.com/mchalla'><img border=0 src='https://avatars0.githubusercontent.com/u/725475?v=3' height=16 width=16></a> Looks good to me.
- <a href='https://github.com/borkmann'><img border=0 src='https://avatars2.githubusercontent.com/u/677393?v=3' height=16 width=16></a> Repushed: fixed help message and added stale global map detection & removal during daemon init.
- [x] <a href='#crh-comment-Pull b3540f1ad5c58546cd169149a9c95467168dbd90 pkg/endpoint/endpoint.go 14'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/cilium/cilium/pull/443#discussion_r107954562'>File: pkg/endpoint/endpoint.go:L83-95</a></b>
- <a href='https://github.com/aanm'><img border=0 src='https://avatars3.githubusercontent.com/u/5714066?v=3' height=16 width=16></a> While I was reading the code I was asking my self "What is local? Do we have a non-local CT?" Now I get it, can I purpose to change the name to `Use container dedicated connection tracking..`? It's just a thought
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> I think we are using the word endpoint pretty much everywhere as Cilium may be extended to non container specific use cases. Let's be consistent.
- <a href='https://github.com/borkmann'><img border=0 src='https://avatars2.githubusercontent.com/u/677393?v=3' height=16 width=16></a> Makes sense, sure.
- <a href='https://github.com/aanm'><img border=0 src='https://avatars3.githubusercontent.com/u/5714066?v=3' height=16 width=16></a> @tgraf oh I was asking the `local` vs `dedicated`. The word container was already there.
- <a href='https://github.com/borkmann'><img border=0 src='https://avatars2.githubusercontent.com/u/677393?v=3' height=16 width=16></a> I'm using `Use endpoint dedicated tracking table instead of global one` now.
- [x] <a href='#crh-comment-Pull b3540f1ad5c58546cd169149a9c95467168dbd90 tests/01-ct.sh 7'></a> <img src='http://www.codereviewhub.com/site/github-completed.png' height=16 width=60>&nbsp;<b><a href='https://github.com/cilium/cilium/pull/443#discussion_r107949091'>File: tests/01-ct.sh:L198-233</a></b>
- <a href='https://github.com/tgraf'><img border=0 src='https://avatars2.githubusercontent.com/u/1417913?v=3' height=16 width=16></a> Agreed. reconfig should remove old stale maps. Cilium should also remove stale maps on startup.
- <a href='https://github.com/borkmann'><img border=0 src='https://avatars2.githubusercontent.com/u/677393?v=3' height=16 width=16></a> Potentially cilium could even do best-effort and migrate existing connection from the global to the local table.


<a href='https://www.codereviewhub.com/cilium/cilium/pull/443?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/cilium/cilium/pull/443?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/cilium/cilium/pull/443'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

tgraf commented on tests/01-ct.sh:
```diff
@@ -198,8 +198,35 @@ function connectivity_test() {
+function cleanup_bpffs() {
+	# Cleanup so we can detect bugs in case these get regenerated when
+	# they shouldn't. We could perhaps let cilium prune stale maps in
+	# future on reconfig?
```
Agreed. reconfig should remove old stale maps. Cilium should also remove stale maps on startup.

borkmann replied:
Potentially cilium could even do best-effort and migrate existing connection from the global to the local table.

tgraf replied:
👍

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -82,6 +83,12 @@ var (
 
+	OptionSpecConntrackLocal = option.Option{
+		Define:      "CONNTRACK_LOCAL",
+		Description: "Use container local connection tracking table instead of global one",
```
While I was reading the code I was asking my self "What is local? Do we have a non-local CT?" Now I get it, can I purpose to change the name to `Use container dedicated connection tracking..`? It's just a thought

tgraf replied:
I think we are using the word endpoint pretty much everywhere as Cilium may be extended to non container specific use cases. Let's be consistent.

borkmann replied:
Makes sense, sure.

aanm replied:
@tgraf oh I was asking the `local` vs `dedicated`. The word container was already there.

borkmann replied:
I'm using `Use endpoint dedicated tracking table instead of global one` now.

mchalla committed:
bpf: Global 5 tuple conntrack.

- support both global 5 tuple / container local 4 tuple conntrack
- cilium daemon can configure this support using #define CONNTRACK_LOCAL
- defaults to global

Signed-off-by: Madhu Challa <madhu@cilium.io>

borkmann committed:
cilium, bpf: make global vs local CT table configurable

Use a single huge global connection tracking table by default, and
have also the option to switch to container local one instead as we
have prior to this patch.

The switch can even happen during runtime per reconfiguration via:

  cilium endpoint config <ep-ID> ConntrackLocal=false  // global table
  cilium endpoint config <ep-ID> ConntrackLocal=true   // local table

Dumping the global table works as usual:

  cilium bpf ct list global    // dump global table
  cilium bpf ct list <ep-ID>   // dump local table

It's also listed as a config option via cli, f.e.:

  # cilium config
  Conntrack                Enabled
  ConntrackAccounting      Enabled
  ConntrackLocal           Enabled
  DropNotification         Enabled
  NAT46                    Enabled
  Policy                   Enabled
  PolicyTracing            Disabled

I changed the 01-ct tests to include local/global table tests and
they were all passing.

Signed-off-by: Daniel Borkmann <daniel@cilium.io>

mchalla committed:
Add global, local ctmaps with different key sizes.

Signed-off-by: Madhu Challa <madhu@cilium.io>

borkmann committed:
cilium: update bindata due to bpf code changes

Update go-bindata due to prior changes under bpf/.

Signed-off-by: Daniel Borkmann <daniel@cilium.io>

borkmann committed:
tests: get rid of bombardier tool/test and fixup ping tests

Never really worked well on IPv6. The ping tests also need
to run with --privileged.

Signed-off-by: Daniel Borkmann <daniel@cilium.io>

borkmann committed:
bpf: remove/cleanup various stuff

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
test extended

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
